### ISSUE TITLE
Support for file download

### ISSUE THREAD

**Issue by b-liberman** [swagger-api/swagger-ui#3149]
#swagger-ui 3.0 does not generate a download link for files.

- swagger-ui: 3.0
- swagger file:
```json
{
  "swagger" : "2.0",
  "info" : {
    "description" : "descr",
    "version" : "0.1",
    "title" : "title"
  },
  "host" : "SM05720",
  "basePath" : "/",
  "schemes" : [ "https" ],
  "paths" : {
    "/test/download" : {
      "get" : {
        "tags" : [ "download" ],
        "description" : "this is to test the download",
        "operationId" : "download XYZ",
        "produces" : [ "application/msexcel" ],
        "parameters" : [ ],
        "responses" : {
          "200" : {
            "description" : "success - download the file",
            "schema" : {
              "type" : "file"
            }
          }
        }
      }
    }
  }
}
```

**Comment by webron** [comment]
@b-liberman do you have a live server we can test? It's a bit difficult to check without an actual server.

**Comment by b-liberman** [comment]
Unfortunately our servers are all behind a firewall.

**Comment by hkosova** [comment]
@webron, try this:
```
swagger: "2.0"
info:
  title: test
  version: "1.0"
    
host: raw.githubusercontent.com
schemes: [https]

paths:
  /mozilla/pdf.js/master/examples/helloworld/helloworld.pdf:
    get:
      produces:
        - application/pdf
      responses:
        '200':
          description: OK
          schema:
            type: file
```

Swagger UI 2.2.10:
![download-file](https://cloud.githubusercontent.com/assets/8576823/26643800/37a7a4e4-463b-11e7-9ae3-f1d702d26dba.png)

**Comment by webron** [comment]
Thanks @hkosova!

**Comment by GreatYYX** [comment]
Hi,
I tried to add `type` in response, but why swagger shows `"Unknown Type: file"`?
```
get:
      tags:
      - "glossaries"
      summary: "Get the content of a glossary"
      description: ""
      operationId: ""
      consumes:
      - "application/json"
      produces:
      - "text/plain; charset=utf-8"
      parameters:
      - in: "path"
        name: "project_name"
        required: true
        description: ""
      - in: "path"
        name: "glossary_name"
        description: ""
        required: true
        description: ""
      responses:
        200:
          description: "ok"
          schema:
            type: "file"
        400:
          description: "bad request"
        404:
          description: "not found"
```

**Comment by webron** [comment]
@GreatYYX I assume you're talking about the generated example. Please file a separate ticket for that.

**Comment by yuanchanghang** [comment]
Xxx

**Comment by pauldzy** [comment]
Hello GreatYYX,

Were you going to open a ticket for the unknown type file issue?  If not I can do so.

Cheers,
Paul

**Comment by eLvErDe** [comment]
Hi,

I can confirm the "example" issue on 3.0.12. Download work as expected but the example says "Unknown Type: file".

Regards, Adam.

**Comment by webron** [comment]
We have a pending PR to resolve the example issue - https://github.com/swagger-api/swagger-ui/pull/3266.

**Comment by eLvErDe** [comment]
In the meanwhile, I defined example with a comment describing filename naming and it "hides" the error message.

**Comment by deepika06** [comment]
Even i am facing the similar problem. The Download API returns with 200 response but i am not able download the file. I see that there is no link to download the file.
I am using latest swagger UI : 3.4.0

**Comment by BanitaChhabra** [comment]
Even i am facing the similar problem. Please guide.
The Download API returns with 200 response but i am not able download the file. I see that there is no link to download the file.
I am using latest swagger UI : 3.4.0

**Comment by heldersepu** [comment]
With the latest I do get a download link:
http://swashbuckletest.azurewebsites.net/swagger/ui/index#/Image/Image_Put

I think this issue can be closed

**Comment by conorlennon** [comment]
The download link is only present when the response has a content-length.
swagger-ui 2 didn't require the content-length to be present in the response.

**Comment by heldersepu** [comment]
@conorlennon That issue with the content-length will be a separate issue...
and what does it mean when a response does not have a content-length?

I still believe this issue can be closed

**Comment by conorlennon** [comment]
Dynamically generated responses often don't have a content-length, particularly chunked responses.

If I remove the content-length check from the ResponseBody render method, the download link is presented for a response without a content-length, and the download link works fine.

**Comment by heldersepu** [comment]
@conorlennon sounds like an easy fix... why don't you send a Pull Request?

**Comment by conorlennon** [comment]
I've just created pull request #3895.
First time doing that, so not sure if I've done everything correctly.
Hopefully I have.

**Comment by BanitaChhabra** [comment]
I get this response with https 
Server response
Code	Details
200	
Response headers

 x-backside-transport: OK OK  connection: Keep-Alive  transfer-encoding: chunked  content-language: en-US  content-type: application/zip  date: Tue, 14 Nov 2017 11:14:12 GMT  p3p: CP="NON CUR OTPi OUR NOR UNI"  x-content-type-options: nosniff  cache-control: no-cache, no-store, must-revalidate  expires: 0  x-xss-protection: 1; mode=block  content-disposition: attachment; filename=kpiexport_all_2017-11-14_2165411535502891649.zip  x-powered-by: Servlet/3.0  content-security-policy: script-src 'self' 'unsafe-eval' 'unsafe-inline'  strict-transport-security: max-age=43000; includeSubDomains; preload  pragma: no-cache  x-client-ip: 172.16.32.10  x-global-transaction-id: 72889088 

And same is perfectly executed with http.

The file gets created at backend, but doe not get link in swagger to see that file
[Doc1.docx](https://github.com/swagger-api/swagger-ui/files/1470514/Doc1.docx)

**Comment by conorlennon** [comment]
@BanitaChhabra that looks like what I would expect - you don't have the content-length (because you ave chunked transfer encoding) in the response, so the download link is not displayed.

**Comment by BanitaChhabra** [comment]
I tried the below option through REST APIi:-
Response.ok(zippedFile)
                .header("Content-Disposition", "attachment; filename=" + zippedFile.getName() )
                .build();

@POST
	@Consumes({ javax.ws.rs.core.MediaType.APPLICATION_JSON })
	@Produces("application/zip")

Are u suggesting to set content length ?

**Comment by conorlennon** [comment]
For now, setting the content-length in the response is required to get the download link.
If pull request #3895 is accepted, then setting the content-length will no longer be required (once the change ends up in a release).

**Comment by shockey** [comment]
OK, #3895 is merged. It _looks_ like the content length problem was the outstanding issue: OP never followed up, and `Unknown type: file` was solved in #3266.

If there's anything outstanding that I've missed, feel free to reply here and I'll reopen.

**Comment by heldersepu** [comment]
@mansoor-ul-haq look at the swagger.json for the details:
http://swashbuckletest.azurewebsites.net/swagger/ui/index#/Image/Image_Put

The theme is one of these: 
https://github.com/ostranme/swagger-ui-themes
Not sure which one you saw, I change them all the time
